shared/fs/container.js

```diff
@@ -11,39 +11,57 @@ type OwnProps = {
 }
 
 type StateProps = {
+  _itemNames: I.List<string>,
+  _username?: string,
+  _pathItems: I.Map<Types.Path, Types.PathItem>,
+  _sortSetting: Types.SortSetting,
+
   path: Types.Path,
-  items: I.List<string>,
   progress: Types.ProgressType,
 }
 
-type DispatchProps = {}
+type DispatchProps = {
+  loadFolderList: (path: Types.Path) => void,
+}
 
 const mapStateToProps = (state: TypedState, {routeProps}: OwnProps) => {
   const path = Types.stringToPath(routeProps.get('path', Constants.defaultPath))
   const itemDetail = state.fs.pathItems.get(path)
-  const items = itemDetail && itemDetail.type === 'folder' ? itemDetail.get('children', I.List()) : I.List()
-  const progress: Types.ProgressType = itemDetail ? itemDetail.progress : 'pending'
-  return {items, path, progress}
+  return {
+    _itemNames: itemDetail && itemDetail.type === 'folder' ? itemDetail.get('children', I.List()) : I.List(),
+    _username: state.config.username || undefined,
+    _pathItems: state.fs.pathItems,
+    _sortSetting: state.fs.pathUserSettings.get(path, Constants.makePathUserSetting()).get('sort'),
+    path,
+    progress: itemDetail ? itemDetail.progress : 'pending',
+  }
 }
 
 const mapDispatchToProps = (dispatch: Dispatch) => ({
-  _loadFolderList: (path: Types.Path) => dispatch(FsGen.createFolderListLoad({path})),
+  loadFolderList: (path: Types.Path) => dispatch(FsGen.createFolderListLoad({path})),
 })
 
-const mergeProps = ({path, items, progress}: StateProps, dispatchProps: DispatchProps, ownProps) => {
-  return {
-    items: items.map(name => Types.pathConcat(path, name)).toArray(),
-    path,
-    progress,
-    ...dispatchProps,
-  }
-}
+const mergeProps = (stateProps: StateProps, dispatchProps: DispatchProps, ownProps) => ({
+  items: Constants.sortPathItems(
+    stateProps._itemNames.map(name => Types.pathConcat(stateProps.path, name)).map(
+      p => stateProps._pathItems.get(p, Constants.makeUnknownPathItem()) // provide an unknown default to make flow happy
+    ),
+    stateProps._sortSetting,
+    Types.pathIsNonTeamTLFList(stateProps.path) ? stateProps._username : undefined
+  )
+    .map(({name}) => Types.pathConcat(stateProps.path, name))
```

Comment:
We're doing this map-concat twice. Can you cache the results and map over those instead?

Reply:
Oh, Nojima already gave a similar comment about caching the results in a `const`.

Reply:
@jzila The best I can think of without introducing a `path` field into `PathItem` is to have a wrapper object that includes a `pathItem: PathItem` field and a `path: Path` field, or have an object that have a `path: Path` field along with all fields in `PathItem`. But either way it seems I have to call `toObject()` on the `PathItem` returned from `get` to make it work, which seems to be more work than concat'ing itself. Let me know if you have a better idea!